### PR TITLE
virtio-devices: watchdog: Handle out-of-bound access

### PR DESCRIPTION
**Pull request by likebreath**
This PR depends on #4626.

Signed-off-by: Bo Chen <chen.bo@intel.com>

### CODE REVIEW

**Comment by likebreath** [comment]
@rbradford I am not sure this is the right fix. But I think this is a legit programming error (out-of-bound) and it should be handled explicitly. Please advise. Thank you.

**Comment by rbradford** [comment]
We have a panic handler in every virtio thread which catches panics and turns them into an error - further the iteration through the loop is bounded by the size of the queue which is QUEUE_SIZE.

The code however does look different to our usual virtio queue handling - so it might be reasonable to make it look similar (i.e. move the `.add_used()` into the loop step)

File: virtio-devices/src/watchdog.rs [hunk 1]
```diff
@@ -96,7 +98,9 @@ impl WatchdogEpollHandler {
                 self.last_ping_time.lock().unwrap().replace(Instant::now());
             }
 
-            used_desc_heads[used_count] = (desc_chain.head_index(), len);
+            *used_desc_heads
+                .get_mut(used_count)
+                .ok_or(Error::UnexpectedAvailDescriptor)? = (desc_chain.head_index(), len);
```
Comment (sboeuf) [inline virtio-devices/src/watchdog.rs:103]: The approach is correct but let's take this as an opportunity to make the watchdog implementation follow the other virtio devices implementations. And with the new design, there will be no out-of-bound access.

Comment (sboeuf) [reply virtio-devices/src/watchdog.rs:103]: I'm gonna send a PR in a sec!

**Pull request closed without merging**